File: gcj_2016/QR_A.hs
-- Google Code Jam 2016 Qualification Round
-- Problem A. Counting Sheep

import Control.Monad
import Data.Char
import Text.Printf

readInt = (read <$> getLine) :: IO Int

check :: Int -> Int -> Bool
check d n = elem d $ map digitToInt $ show n

findFirst :: Int -> Int -> Int
findFirst d n = head $ filter (\x -> check d x) [n, n*2..]

solve :: Int -> String
solve n = if n == 0 then "INSOMNIA" else show $ maximum $ map (\d -> findFirst d n) [0..9]

main :: IO ()
main = do
  t <- readInt
  forM_ [1..t] $ \i -> do
    n <- readInt
    printf "Case #%d: %s\n" i $ solve n

File: gcj_2016/QR_A.cpp
// Google Code Jam 2016 Qualification Round
// Problem A. Counting Sheep

#include <iostream>
#include <sstream>

using namespace std;

typedef long long LL;

LL solve(LL N) {
	LL f = 0, x;
	for (x = N; ; x += N) {
		char temp[16];
		sprintf(temp, "%lld", x);
		for (int i = 0; temp[i]; ++i) {
			f |= (1 << (temp[i] - '0'));
		}
		if (f >= 1023) {
			break;
		}
	}
	return x;
}

int main(int argc, char *argv[]) {
	LL T;
	cin >> T;
	for (int t = 1; t <= T; ++t) {
		LL N;
		cin >> N;
		if (N == 0) {
			cout << "Case #" << t << ": INSOMNIA" << endl;
		} else {
			cout << "Case #" << t << ": " << solve(N) << endl;
		}
	}
	return 0;
}
